File: Java/longest_substring.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        /*

        with out dups each char occurs only once

        brute 

        iterate for each char from its next and break when same char occurs 

        two pointer approch // have a hash map to store the index of the charecer to figure out the length
        two pointer a and b, b will keep moving 

        */

        HashSet<Character> hset = new HashSet<>();
        int left = 0;
        int res = 0; 

        for(int right = 0; right<s.length(); right++){

            // char is dup
            // update window and check window
            while(hset.contains(s.charAt(right))){
                // update window and remove dup element
                hset.remove(s.charAt(left));
                left+=1;
            }
            
            hset.add(s.charAt(right));
            res = Math.max(res, right-left+1);
        }

        return res;

    }
}

File: Python/Strings/longest_substring.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        """
        approch :
        sliding window  have two pointer to indicate starting and ending of the substring.
        a dict to start the charecter and index of it as we iterate through the string

        """

        last_seen = set()
        left =0 # starting of the substring
        res = 0

        for right, char in enumerate(s):
            while char in last_seen:
                # we found an occurence so we can move on next window 
                # remove/update the begining of the window 
                # while loop to check the rest the current window
                last_seen.remove(s[left])
                left += 1
            #new char caluclate length
            last_seen.add(char)
            res = max(res, right-left+1)
            
        return res

        